Add reroute processor

felixbarny opened the pull request:
[Docs preview](https://elasticsearch_76511.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/reroute-processor.html)

The reroute processor allows to route a document to another target index or data stream. It has two main modes:

When setting the destination option, the target is explicitly specified and the dataset and namespace options can’t be set.

When the destination option is not set, this processor is in a data stream mode. Note that in this mode, the reroute processor can only be used on data streams that follow the [data stream naming scheme](https://elasticsearch_76511.docs-preview.app.elstc.co/guide/en/fleet/master/data-streams.html#data-streams-naming-scheme). Trying to use this processor on a data stream with a non-compliant name will raise an exception.

The name of a data stream consists of three parts: <type>-<dataset>-<namespace>. See the [data stream naming scheme](https://elasticsearch_76511.docs-preview.app.elstc.co/guide/en/fleet/master/data-streams.html#data-streams-naming-scheme) documentation for more details.

This processor can use both static values or reference fields from the document to determine the dataset and namespace components of the new target. See [Table 38, “Reroute options”](https://elasticsearch_76511.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/reroute-processor.html#reroute-options) for more details.

It’s not possible to change the type of the data stream with the reroute processor.
After a reroute processor has been executed, all the other processors of the current pipeline are skipped, including the final pipeline. If the current pipeline is executed in the context of a [Pipeline](https://elasticsearch_76511.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/pipeline-processor.html), the calling pipeline will be skipped, too. This means that at most one reroute processor is ever executed within a pipeline, allowing to define mutually exclusive routing conditions, similar to a if, else-if, else-if, … condition.

The reroute processor ensures that the data_stream.<type|dataset|namespace> fields are set according to the new target. If the document contains a event.dataset value, it will be updated to reflect the same value as data_stream.dataset.

Note that the client needs to have permissions to the final target. Otherwise, the document will be rejected with a security exception which looks like this:

```
{"type":"security_exception","reason":"action [indices:admin/auto_create] is unauthorized for API key id [8-dt9H8BqGblnY2uSI--] of user [elastic/fleet-server] on indices [logs-foo-default], this action is granted by the index privileges [auto_configure,create_index,manage,all]"}
```

Reroute options

| Name        | Required  | Default  | Description |
|-|-|-|-|
| `destination` | no        | -        | A static value for the target. Can't be set when the `dataset` or `namespace` option is set.
| `dataset`     | no        | `{{data_stream.dataset}}`   | Field references or a static value for the dataset part of the data stream name. In addition to the criteria for <<indices-create-api-path-params, index names>>, cannot contain `-` and must be no longer than 100 characters. Example values are `nginx.access` and `nginx.error`. <br><br>Supports field references with a mustache-like syntax (denoted as `{{double}}` or `{{{triple}}}` curly braces). When resolving field references, the processor replaces invalid characters with `_`. Uses the `<dataset>` part of the index name as a fallback if all field references resolve to a `null`, missing, or non-string value.
| `namespace`   | no        | `{{data_stream.namespace}}` | Field references or a static value for the namespace part of the data stream name. See the criteria for <<indices-create-api-path-params, index names>> for allowed characters. Must be no longer than 100 characters.<br><br>Supports field references with a mustache-like syntax (denoted as `{{double}}` or `{{{triple}}}` curly braces). When resolving field references, the processor replaces invalid characters with `_`. Uses the `<namespace>` part of the index name as a fallback if all field references resolve to a `null`, missing, or non-string value.


The `if` option can be used to define the condition in which the document should be rerouted to a new target.

```json
{
  "data_stream_router": {
    "tag": "nginx",
    "if" : "ctx?.log?.file?.path?.contains('nginx')",
    "type": "logs",
    "dataset": "nginx"
  }
}
```

The dataset and namespace options can contain either a single value or a list of values that are used as a fallback.
If a field reference evaluates to `null`, is not present in the document, the next value or field reference is used.
If a field reference evaluates to a non-`String` value, the processor fails.

In the following example, the processor would first try to resolve the value for the `service.name` field to determine the value for `dataset`.
If that field resolves to `null`, is missing, or is a non-string value, it would try the next element in the list.
In this case, this is the static value `"generic`".
The `namespace` option is configured with just a single static value.


```json
{
  "reroute": {
    "dataset": [
        "{{service.name}}",
        "generic"
    ],
    "namespace": "default"
  }
}
```


Depends on
* https://github.com/elastic/elasticsearch/pull/94000

ruflin commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DataStreamRouterProcessor.java:
```diff
@@ -0,0 +1,149 @@
+    private static final String DATA_STREAM_DEFAULT_NAMESPACE = "default";
+    private static final String EVENT_DATASET = "event.dataset";
+
+    private static final char[] DISALLOWED_IN_DATASET = new char[] { '\\', '/', '*', '?', '\"', '<', '>', '|', ' ', ',', '#', ':', '-' };
```
Can you make sure this is aligned with what we have in the https://github.com/elastic/package-spec for validation? @mtojek Will likely know where to point you to. Same for namespace.

ruflin replied:
@mtojek @jsoriano Can you chime in here? I would have expected that we do some validation.

jsoriano replied:
We have this issue open since long ago https://github.com/elastic/package-spec/issues/57, I will ensure this ends up in our backlog.

ruflin commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DataStreamRouterProcessor.java:
```diff
@@ -0,0 +1,149 @@
+        s = s.toLowerCase(Locale.ROOT);
+        s = s.substring(0, Math.min(s.length(), MAX_LENGTH));
+        for (char c : disallowedInDataset) {
+            s = s.replace(c, REPLACEMENT_CHAR);
```
I don't like the magical replacement. Ca we just make it invalid and return an error?

felixbarny replied:
When you configure the processor with an invalid value, it will return an error.

However, when taking values from the document, it will be more lenient. One use case for the leniency is the fallback to `event.dataset`. ECS logger have first-class support to setting the value as it's an important field for the logs UI which shows log rates per `event.dataset`. The `event.dataset` field did not have restrictions on the value previously. Introducing them retrospectively would be a breaking change. Therefore, I opted to sanitize rather than reject invalid values. See also the [ECS RFC about data streams](https://github.com/elastic/ecs/blob/main/rfcs/text/0009-data_stream-fields.md#scope-of-impact):

> Since data_stream.dataset and event.dataset should contain the same value, the restrictions imposed on data_stream.dataset might affect the event.dataset value. This means users may need to translate their custom dataset values (e.g. event.dataset: firewall/config) to an equivalent legal dataset, according to the character restrictions imposed by the use of the value in data_stream.dataset, for example data_stream.dataset: firewall.config.

We could change it so that only `event.dataset` values will be sanitized. But completely dropping an event if the target contains an invalid character seems somewhat harsh, especially since we don't have a dead letter queue.
Do we really prefer to completely drop a log event if it contains `"data_stream.dataset": "my-service"`? To me, sanitizing the value from `my-service` to `my_service` seems more user-friendly.

But I'm open to changing this if you still feel strongly about it.

felixbarny replied:
> So building event.dataset into more places will make it even harder to get rid off. Could we skip event.dataset as a fallback in a first iteration?

As it is already built in into ECS loggers and the logs UI, not having support for it would cause challenges. But we could add a flag to the processor (such as `event_dataset_fallback`) that's off by default.

ruflin replied:
++

felixbarny replied:
Removed any usage of `event.dataset`, as discussed. If fallbacks are desired, a `set` processor can be used.

ruflin commented on docs/reference/ingest/processors/data-stream-router.asciidoc:
```diff
@@ -0,0 +1,40 @@
+See the {fleet-guide}/data-streams.html#data-streams-naming-scheme[data stream naming scheme] documentation for more details.
+
+After a `data_stream_router` processor has been executed, all the other processors of the current pipeline are skipped.
+This means that at most one `data_stream_router` processor is ever executed within a pipeline,
```
If the data is then delivered to the next data stream, will it be possible to have their another data_stream_router?

Example:

* We detect it is an nginx image, route to `logs-nginx-default`
* On this data stream, there is another data stream router to decide if it is access or error logs. If none match, it is stores in `logs-nginx-default`, otherwise sent to `logs-nginx.access-default` where there is likely and ingest processor to take the message apart.

I remember there were historically some problems around this.

felixbarny replied:
Yes, this should be possible. But I didn't explicitly test that scenario yet. I'll make sure to add tests for that if we decide to move forward with this approach.

felixbarny replied:
I've added automated tests for this scenario. See `300_data_stream_processor.yml`.

felixbarny replied:
Turned out there's an issue with that. 

https://github.com/elastic/elasticsearch/pull/59522
> default pipeline from new index won't be executed

So currently, multi-stage routing is not working. It seems that's due to this concern mentioned in https://github.com/elastic/elasticsearch/issues/57968
> how do we prevent the document from being infinitely redirected between indices

I took a stab at allowing default pipelines of the new index to be executed in [f7c2c51](https://github.com/elastic/elasticsearch/pull/76511/commits/f7c2c510fa0e572b17b637889499c2ada25dd6d3).

felixbarny replied:
I've extracted this to a separate PR:
* https://github.com/elastic/elasticsearch/pull/85931

felixbarny commented:
Discussing some option questions:

### Only set `event.dataset` if it is already set on the document to ensure that it's the exact same value as `data_stream.dataset`?
This has the implication that ECS loggers need to set `event.dataset` in addition or instead of `data_stream.dataset` so that the logs rate component in the logs ui works properly. But to be compatible with the widest range of stack versions and UI features that rely on `event.dataset`, ECS loggers will need to set both anyway. See also https://github.com/elastic/ecs-logging/pull/38

Overall a net-positive change in my books to not overly promote `event.dataset`. Implemented in c30141d. 

### Don't treat `event.dataset` as a fallback for routing if `data_stream.dataset` is empty?
This doesn't need to be part of this processor. If this makes sense for specific use cases (I'm thinking about ECS logs), we can just add a processor like this:

```yaml
- set:
    field: data_stream.dataset
    copy_from: event.dataset
    if: ctx.event?.dataset instanceof String && ctx.event.dataset.length() > 1
    override: false
```

If we can rely on the fact that shippers always populate `data_stream.*` fields when sending events to a routing pipeline, we could just ignore `event.dataset` and simplify the logic.

There's a special case for ECS loggers where the shipper would send the data to a generic routing pipeline but


Other than having the processor natively falling back to `event.dataset`, users could just use a `set` processor if they want the fallback.

However, this has the implication that if `event.dataset: foo` is set on the document, and `data_stream.dataset` is unset, the doc will be routed to the `generic` dataset and `event.dataset` will be overridden to `generic`. That's to comply with the spec that `event.dataset` and `data_stream.dataset` should be the same.

This is quite a constructed case, however, as I can't think of a valid case where `event.dataset` is set and `data_stream.dataset` is missing. That's even true for ECS loggers. While the might set only the former, the log message will end up as a JSON-encoded string in the `message` field. We can make a fallback in the specific ECS JSON log pipeline.

### Should we sanitize `data_stream.*` fields that contain invalid characters?

I think we should. This makes it much easier to route events to a data stream with dynamically computed values. For example using [`azure.springcloudlogs.app_name`](https://github.com/elastic/integrations/blob/f2901e29d9d729bc054f92c49c9d428cf972c4a9/packages/azure/data_stream/springcloudlogs/sample_event.json#L53) as the dataset. There's no restrictions on the allowed characters for this field. Still, it makes for a good dataset. Rather than each processor having to sanitize dynamic values, it seems better to rely on the data_stream_router processor for sanitization.

### Should we parse `_index` if `data_stream.*` fields are missing?
While our integrations set `data_stream.*` fields, there's no general guarantee that these are available. Especially when events are coming from shippers we don't control, these might be missing.

It seems sensible for this processor to guarantee that the index name is always consistent with the `data_stream.*` fields. It could do so by falling back to parsing `_index` rather than always using `dataset: generic` `namespace: default` as a fallback.

But I'd say this is an optional enhancement of the processor that's not necessarily in scope of the first iteration.

felixbarny commented:
> Should we parse _index if data_stream.* fields are missing?

Implemented in 41c23cc.

Implications:
- Processor can only be used on data streams matching `<type>-<dataset>-<namespace>`
- Ensures data_stream.* fields are in sync with _index, even if the source doc didn't contain data_stream.* fields
- Disallow changing type, only inferred from _index
- The type is not configurable in on the processor

felixbarny committed:
Add data_stream_router processor

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

ruflin commented:
@felixbarny My general comment around https://github.com/elastic/elasticsearch/pull/76511#issuecomment-1099098786 is, lets keep the first iteration as simple as possible. It is always possible to add more features.

* Lets ignore event.dataset, user can use set processor
* Sanitise: Can we skip it at first or make it a config option that is off by default? I agree that we will need it eventually.
* Parse _index: Would be great, but could be done as an enhancement

felixbarny commented:
> Lets ignore event.dataset, user can use set processor

Currently, the processor only makes sure that if the source event contains `event.dataset` it doesn't conflict with `data_stream.dataset`. But I agree that this could also be done with a set processor. However, the risk of that is that `event.dataset` and `data_stream.dataset` may be inconsistent.

> Sanitise: Can we skip it at first or make it a config option that is off by default? I agree that we will need it eventually.

What are the cases where you'd want to have sanitization off? Could we make sanitization on by default (which is what the PR currently does) and later add a switch to disable it?

> Parse _index: Would be great, but could be done as an enhancement

I've added that already. At the same time, I've removed the ability to set `data_stream.type` in the processor as it will always use the type derived from the index name. Hence, I think this should go in with the first version as it wouldn't be possible to add this later without breaking changes.

ruflin commented:
> What are the cases where you'd want to have sanitization off? Could we make sanitization on by default (which is what the PR currently does) and later add a switch to disable it?

If it does the sanitization, will it also cleanup the document itself meaning modifying data_stream.dataset value? My suggestion having it off by default is mostly my instinct to disable magic by default which requires users to think about why they do not match the default. But I see your scenario that users just want to ingest logs and should not be exposed to this detail. I'm on board with turning it on by default which could also mean not having the config at first ;-)

felixbarny commented:
This is how the different components of the data stream name (`<type>-<dataset>-<namespace>`) are determined: 

| | processor config | data_stream.* value from the document | _index | description
|-|-|-|-|-|
| type| 🚫 |🚫 | ✅ | The type is always taken from the index name and can't be changed via the processor config or by overriding the `data_stream.type` field in the document.
| dataset | ✅| ✅🧼  | ✅| The processor can be configured with a static `dataset` value that will take precedence over values in the document and the index name. If not set, the sanitized `data_stream.dataset` value from the document is used. This allows pipelines to route the event to another data stream by setting `data_stream.dataset` to a custom value, without having to worry about which characters are allowed. In the edge case where `data_stream.dataset` is missing from the document, the processor will figure it out from the index name (the part in-between the first and the second dash)
| namespace | ✅| ✅🧼  | ✅| The processor can be configured with a static `namespace` value that will take precedence over values in the document and the index name. If not set, the sanitized `data_stream.namespace` value from the document is used. This allows pipelines to route the event to another data stream by setting `data_stream.namespace` to a custom value, without having to worry about which characters are allowed. In the edge case where `data_stream.namespace` is missing from the document, the processor will figure it out from the index name (the part after the second dash)

After the three components of the data stream name have been determined, the processor explicitly sets and overrides the corresponding `data_stream.<type|dataset|namespace>` values in the document. Additionally, it overrides `_index` to ensure the value is consistent with the fields it has set on the document (`<type>-<dataset>-<namespace>`).

The processor will throw an exception if the `_index` does not conform to the `<type>-<dataset>-<namespace>` naming scheme.

felixbarny committed:
Don't override event.dataset

felixbarny committed:
Merge remote-tracking branch 'origin/main' into data_stream_router_processor

elasticsearchmachine commented:
Hi @felixbarny, I've created a changelog YAML for you.

felixbarny committed:
Update docs/changelog/76511.yaml

ruflin commented on docs/reference/ingest/processors/data-stream-router.asciidoc:
```diff
@@ -0,0 +1,43 @@
+The name of a data stream is comprised of three parts and looks like this: `<type>-<dataset>-<namespace>`.
+See the {fleet-guide}/data-streams.html#data-streams-naming-scheme[data stream naming scheme] documentation for more details.
+
+NOTE: `data_stream_router` processor can only be used on data streams that follow the data streams naming scheme.
```
We should add the link to https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme here

ruflin replied:
No. There are some mentions of it in the docs but it also links to the blog post.

felixbarny replied:
Seems like we do have docs an I've linked it on line 11: https://www.elastic.co/guide/en/fleet/current/data-streams.html#data-streams-naming-scheme

ruflin commented on docs/reference/ingest/processors/data-stream-router.asciidoc:
```diff
@@ -0,0 +1,43 @@
+{
+  "data_stream_router": {
+    "tag": "nginx",
+    "if" : "ctx?.log?.file?.path?.contains('nginx')",
```
I remember there were some efforts to simplify this. Do we still need `ctx`? Would be nice if users could just reference a field, no matter where it is stored.

Same for all the `?.`, users should not have to care and just type it in. I know, more an underlying ES issue but maybe we already work around it in the code of this processor.

ruflin commented on docs/reference/ingest/processors/data-stream-router.asciidoc:
```diff
@@ -0,0 +1,43 @@
+|======
+
+NOTE: It's not possible to change the `type` of the data stream by setting the `data_stream.type` in the document.
+
```
What will happen if the user does not have permissions for the target index? We should verify that a good error is returned for the rejection (I assume doc is rejected)

felixbarny replied:
You'll get the same error as trying to directly index into a data stream you don't have permissions for. This is how the error would look like:

```json
{"type":"security_exception","reason":"action [indices:admin/auto_create] is unauthorized for API key id [8-dt9H8BqGblnY2uSI--] of user [elastic/fleet-server] on indices [logs-foo-default], this action is granted by the index privileges [auto_configure,create_index,manage,all]"}
```

ruflin commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DataStreamRouterProcessor.java:
```diff
@@ -0,0 +1,143 @@
+        if (indexOfFirstDash < 0) {
+            throw new IllegalArgumentException(illegalDataStreamNameMessage);
+        }
+        type = indexName.substring(0, indexOfFirstDash);
```
Lets extract this to a separate function.

felixbarny replied:
I've extracted some methods but I'm not sure if it's cleaner now. Having support for multiple return values, like in Go, would come in handy for this.

felixbarny committed:
Merge branch 'main' into data_stream_router_processor

felixbarny committed:
Use IngestDocument#redirect

felixbarny committed:
Extract data stream parsing methods

felixbarny committed:
Add changelog

felixbarny committed:
Revert "Add changelog"

This reverts commit ac5f282b61376c4a4a7383e2c4683ce2a38a9aa6.

andrewkroh commented on docs/reference/ingest/processors/data-stream-router.asciidoc:
```diff
@@ -0,0 +1,43 @@
+This means that at most one `data_stream_router` processor is ever executed within a pipeline,
+allowing to define mutually exclusive routing conditions,
+similar to a if, else-if, else-if, … condition.
+
```
It might be worth mentioning that "caller" must have permissions to write to the final target data stream.

felixbarny replied:
done

felixbarny committed:
Merge remote-tracking branch 'origin/main' into data_stream_router_processor

felixbarny committed:
Rename processor to reroute and add destination option

felixbarny committed:
Fix docs build

felixbarny committed:
Add support for field references and multiple fallbacks

felixbarny committed:
Add skip_if_target_unchanged option

felixbarny committed:
Use mustache scripts instead of custom field reference syntax

felixbarny committed:
Merge remote-tracking branch 'origin/main' into data_stream_router_processor

felixbarny committed:
Revert "Use mustache scripts instead of custom field reference syntax"

This reverts commit 26b3c793416c9cf9b8a266f5e5226e2a111199dc.

felixbarny committed:
Set event.dataset to be consistent with data_stream.dataset

felixbarny committed:
Revert "Add skip_if_target_unchanged option"

This reverts commit 4888843c845983811a7aee345efedf322ba26994.

felixbarny committed:
Field references to use same syntax as mustache templates

felixbarny committed:
Add comments to RerouteProcessor

felixbarny committed:
Use {{data_stream.dataset}} and {{data_sream.namespace}} as default values

felixbarny committed:
Update docs

felixbarny committed:
Merge remote-tracking branch 'origin/main' into data_stream_router_processor

felixbarny committed:
Apply spotless suggestions

felixbarny committed:
Merge branch 'main' into data_stream_router_processor

ruflin commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RerouteProcessor.java:
```diff
@@ -0,0 +1,274 @@
+                if (fieldReference.startsWith("{") && fieldReference.endsWith("}")) {
+                    fieldReference = fieldReference.substring(1, fieldReference.length() - 1);
+                }
+                // only a single field reference is allowed
```
Does this mean, the user can't set a reference for dataset AND namespace? Maybe worth adding to the docs

felixbarny replied:
What it means is that you can't do `"dataset": "{{foo}}-{{bar}}"`, just `"dataset": "{{foo}}"`

ruflin replied:
Ah, got it. I got confused by the `-`

ruflin reviewed (commented):
I left some comments for the docs. But this should not hold this PR back. We can always follow up with more detailed docs.

What if a users users `type: foo` and then the data stream values? It seems we don't validate the type anywhere? It should only be los, metrics, traces or synthetics.

felixbarny committed:
Update docs/reference/ingest/processors/reroute.asciidoc

Co-authored-by: Nicolas Ruflin <spam@ruflin.com>

joegallo committed:
Tidy up imports

joegallo committed:
Prefer toList

joegallo committed:
Drop a utility method, add a test

joegallo committed:
Add a comment

joegallo committed:
Drop toString

joegallo committed:
Merge branch 'main' into data_stream_router_processor

elasticmachine committed:
Merge branch 'main' into data_stream_router_processor

felixbarny committed:
Add sanitization tests

ruflin commented on docs/reference/ingest/processors/reroute.asciidoc:
```diff
@@ -0,0 +1,93 @@
+// NOTCONSOLE
+
+The dataset and namespace options can contain either a single value or a list of values that are used as a fallback.
+If a field reference evaluates to `null`, is not present in the document, or if the value is not a `String`, the next value or field reference is used.
```
I'm strongly in favor of not dropping data. The user setting up the rule is not in control of the data that is shipped. It can happen that most events are fine but there are some events which have malformed data like by accident an array. 

Ideally, there is an easy way for users to debug this or even see it in the docs that got ingested. That is where I would focus the discussion but we can still follow up on this one.

We can document it as a best practice to have a string without variables inside as last fallback option if the fallback option is not the target data stream.

felixbarny replied:
> We can document it as a best practice to have a string without variables inside as last fallback option if the fallback option is not the target data stream.

We can, but there's always the implicit fallback of getting the value from the current target (`_index: <type>-<dataset>-<namespace>`)

dakrone replied:
I think trying to treat this with leniency is taking some of the data stream failure store functionality and putting it into this reroute processor. I would rather that we throw and exception and fail in that scenario. We can then handle this in a better way once we have that functionality.

Imagine N months from now when we do have a failure store, this leniency means that tracking the location of the data down to where it failed is now harder, because we silently skip a service name parameter with no way to make it "re-strict". If we are strict now, we could always decide at a later time to relax that strictness and allow more leniency, however, we can't go the other direction.

We already have this same sort of strictness in our documents by making things `constant_keyword` (for example, if someone were to write a doc where `data_stream.dataset` was an array).

felixbarny replied:
We've discussed this today and decided to change it to be more strict, as suggested by Joe.

felixbarny committed:
Move sanitization constants

felixbarny committed:
Use regex for sanitization

felixbarny committed:
Merge remote-tracking branch 'origin/main' into data_stream_router_processor

felixbarny committed:
Be more strict regarding non-string field references

felixbarny committed:
Merge remote-tracking branch 'origin/main' into data_stream_router_processor

joegallo reviewed (approved):
🚀
